cpmpy/solvers/pumpkin.py

```diff
@@ -0,0 +1,531 @@
+            
+            elif cpm_expr.name == "cumulative":
+                start, dur, end, demand, cap = cpm_expr.args
+                assert all(is_num(d) for d in dur), "Pumpkin only accepts Cumulative with fixed durations"
+                assert all(is_num(d) for d in demand), "Pumpkin only accepts Cumulative with fixed demand"
+                assert is_num(cap), "Pumpkin only accepts Cumulative with fixed capacity"
+
+                return [constraints.Cumulative(self.solver_vars(start),dur, demand, cap)] + \
+                        [self._get_constraint(c)[0] for c in self.transform([s + d == e for s,d,e in zip(start, dur, end)])]
+            else:
+                raise NotImplementedError(f"Unknown global constraint {cpm_expr}")
+
+        elif isinstance(cpm_expr, BoolVal): # unlikely base case
+            a = boolvar() # dummy variable
+            if cpm_expr.value() is True:
+                return self._get_constraint(Operator("sum", [a]) >= 1)
+            else:
+                return self._get_constraint(Operator("sum", [a]) <= -1)
+
+        else:
+            raise ValueError("Unexpected constraint:", cpm_expr)
+
+
+    def add(self, cpm_expr_orig):
+        """
+            Eagerly add a constraint to the underlying solver.
+
+            Any CPMpy expression given is immediately transformed (through `transform()`)
+            and then posted to the solver in this function.
+
+            This can raise 'NotImplementedError' for any constraint not supported after transformation
+
+            The variables used in expressions given to add are stored as 'user variables'. Those are the only ones
+            the user knows and cares about (and will be populated with a value after solve). All other variables
+            are auxiliary variables created by transformations.
+
+        :param cpm_expr: CPMpy expression, or list thereof
+        :type cpm_expr: Expression or list of Expression
+
+        :return: self
+        """
+        from pumpkin_py import constraints
+
+        # add new user vars to the set
+        get_variables(cpm_expr_orig, collect=self.user_vars)
+
+        # transform and post the constraints
+        for orig_expr in toplevel_list(cpm_expr_orig, merge_and=True):
+            for cpm_expr in self.transform(orig_expr):
+                if isinstance(cpm_expr, Operator) and cpm_expr.name == "->": # found implication
+                    bv, subexpr = cpm_expr.args
+                    for cons in self._get_constraint(subexpr):
+                        if isinstance(cons, constraints.Clause):    
+                            raise ValueError("_get_constraint should not return clauses")
+                        self.pum_solver.add_implication(cons, self.solver_var(bv))
+                else:
+                    solver_constraints = self._get_constraint(cpm_expr)
+                    for cons in solver_constraints:
```

Comment:
do we care if it returns a clause?

Reply:
If we want to tag the constraints we do; but not in the main CPMpy branch indeed.